src/dlsnode/storage/BucketFile.d

```diff
@@ -166,6 +170,132 @@ public class BucketFile: OutputStream
+
+    private ErrnoException errno_exception;
+
+
+    /***************************************************************************
+
+        "Promise" for the scheduled non-blocking async IO operation.
+        This will be set with a result when the async IO operation
+        finishes.
+
+    ****************************************************************************/
+
+    public struct Promise
+    {
+        /***********************************************************************
+
+            Error code returned by AsyncIO system
+
+        ***********************************************************************/
+
+        private int last_errno;
+
+
+        /***********************************************************************
+
+            Number of bytes read by the last non-request-blocking aio request.
+
+        ***********************************************************************/
+
+        private ssize_t bytes_last_read;
+
+
+        /**********************************************************************
+
+            Buffer with the data read.
+
+        **********************************************************************/
+
+        private void[] data_buffer;
+
+
+        /**********************************************************************
+
+            Current state of the promise.
+
+        **********************************************************************/
+
+        private enum State
+        {
+            Empty,
+            PromiseMade,
+            PromiseFulfilled,
+            PromiseReaped
+        }
+
+        /// ditto
+        private State state;
+
+        /**********************************************************************
+
+            Resets the promise to the initial state.
+
+        **********************************************************************/
+
+        private void reset ()
+        {
+            this.state = State.Empty;
+            this.last_errno = 0;
+            this.bytes_last_read = -1;
+        }
+
+
+        /**********************************************************************
+
+            Reaps the promise's result and empties the promise.
+
+        **********************************************************************/
+
+        public void[] result ()
+        {
+            verify(this.state == State.PromiseFulfilled);
+            this.state = State.PromiseReaped;
+
+            return this.data_buffer[0..this.bytes_last_read];
+        }
+
+        /**********************************************************************
+
+            Check if the promise is made and we should wait until it's
+            fulfilled.
+
+        **********************************************************************/
+
+        public bool promise_made ()
+        {
+            return this.state == State.PromiseMade;
+        }
+
+        /**********************************************************************
+
+            Check if there was any error.
```

Comment:
Should the description mention that this can only return true if the promise has been fulfilled?